### DISCUSSION TITLE
What i've missed if my filesystem collection returns nothing?

### DISCUSSION THREAD

**Discussion by HiRolla** [joomlatools/joomlatools-pages#882] [Q&A] [answered]
Hello, Johan! 
I really appreciate your advice.
I have `joomlatools-pages/data/wiki` folder with two markdown files inside
```
Interface language.md
start here.md
```
Also I have `joomlatools-pages/pages/wiki.html.php` containing
```
---
title: Test wiki collection of markdown files
summary: Here I learn how to deal with MD data collections.
collection:
   model: filesystem?path=data://wiki
---
<h2><?= $title ?></h2>
<? echo count(collection()) ?>
<? foreach(collection() as $wikiArticle): ?>
   <p>Title: <?= $wikiArticle->title ?></p>
   <? endforeach; ?>
```
But when I browse `\wiki` page I see that `count(collection())` returns 0, e.g. collection is empty.

So, what I've missed? I've read the whole documentation. Sure my understanding now is just like I am a newbie, but here I tried to combine two of your examples: [Filesystem-Collection](https://github.com/joomlatools/joomlatools-pages/wiki/Filesystem-Collection) and [Cookbook-Articles](https://github.com/joomlatools/joomlatools-pages/wiki/Cookbook-Articles).

On the other hand I get expected 2 titles output if I replace the code above by this one:
```
<? $dt = data('wiki') ?>
<? echo count($dt);?>
<? foreach($dt as $wikiArticle): ?>
   <p>Title: <?= $wikiArticle->summary ?></p>
   <? endforeach; ?>
```
But I can't use this (2-nd) way because then I miss filenames that I need later to transliterate and set up routes somehow for each particular article.

**Comment by johanjanssens** [comment, marked as answer]
The Filesystem collection is build to work with a single file containing structured data. For example; a json file, or an ini file. It's not build to work on a folder. The term `filesystem` denotes the data source, the data comes from the filesystem...

The Data API and the Collection API are two different API's that are build for different purposes.  The Data API is primarily intended to work with static structured data files. Often there are smaller pieces of unrelated data you need to render on a site that doesn't belong anywhere, you could hardcode that data in the html but that makes it hard to change them, so thats where the Data API comes in. You should however not use this API for your main content, that's what collections are for.

You could solve the problem in 2 ways:

1. Build a custom Data collection that integrates with the Data API. Something I haven't done yet, but it does make sense.

2. Build an Obsidian extension for Pages that includes an Obsidian collection, would work similar to the Pages collection but be specifically build for Obsidian.

For your use-case 2 makes the most sense, since you also need to integrate additional features like the handling of the slugs based on the filenames etc.

To get an idea how to build a Pages extension you can check: https://github.com/joomlatools/joomlatools-pages/tree/master/contrib/extensions

**Reply by HiRolla** [reply]
> For your use-case 2 makes the most sense

Yes, I agree! 
I've explored that two Pages extensions. It's very interesting for me! Is there some explanations on extentions structure?

Now I'm doing some simple MVP like `publish-obsidian-files.html.php` wich also lets me learn the php coding and check my ideas on dealing with Obsidian flavored MD features like:

- [x]  internal [[link]]
- [x]  [[link|link without path with this label]] 
- [x]  [[subpath/filename]]
- [x]  Auto locate link target
- [x]  Auto locate used images and other attachments
- [ ]  [[link#heading]] 

Also there should be 
- [ ]  Re-publish only the changed and updated markdown files
- [ ]  Some frontmatter processing like `draft: 1` or `published: 0` and others
- [ ]  Some automatic Map of content or index pages for some presenting of published pages and folder categories
- [ ]  Provide some ?conrol=parameters for the `publish-obsidian-files` URL


This is my first Pages php page that works already. What is the workflow?

1.  Maually or via Syncthing copy some markdown files or branches to any `joomlatools-pages/data/[subfolder]` 
2.  Visit the URL `publish-obsidian-files` to run the publishing script
3.  Use published pages at `example.com/[subfolder]/pagename` or `example.com/[subfolder]/[subpath]/pagename`

This script already does the following:

1. Сreate the appropriate `joomlatools-pages/pages/subfolder`
2. Transliterate, URLify and add `.html` to the filename. So `Start here!.md` becomes `start-here.html.md`
3. Process linking syntax and target locatin
4. Place the processed contents into appropriate`joomlatools-pages/pages/[subfolder]/[subpath]`

I've tested this on my markdown files and on the Obsidian help wiki. Some obsidian files rise Pages exeptions, but the others are displayed pretty cool. Thank you for your really helpfull advises. Yes, by now I got some hardcode mvp, but.. Hope later we can move to the Obsidian Pages extension with the all Pages beauty, simplicity and flexibility!
Best regards, 
Alex

**Reply by johanjanssens** [reply]
@HiRolla Hi Alex, been working on this use-case and implemented a new data collection that makes it very easy to create a collection from data files. You can find the PR here: https://github.com/joomlatools/joomlatools-pages/pull/919 Code is due for 0.23.

Also note that the filesystem collection has been renamed to file to make it easier to understand what it does: https://github.com/joomlatools/joomlatools-pages/pull/909

This should make it easier for you to integrate Obsidian into Pages

Happy coding!
